pbsmetrics/prometheus/prometheus.go

```diff
@@ -0,0 +1,242 @@
+package prometheusmetrics
+
+import (
+	"time"
+
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prebid/prebid-server/pbsmetrics"
+	"github.com/prometheus/client_golang/prometheus"
+	_ "github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// Defines the actual Prometheus metrics we will be using. Satisfies interface MetricsEngine
+type Metrics struct {
+	Registry      *prometheus.Registry
+	connCounter   prometheus.Gauge
+	connError     *prometheus.CounterVec
+	imps          *prometheus.CounterVec
+	requests      *prometheus.CounterVec
+	reqTimer      *prometheus.HistogramVec
+	adaptRequests *prometheus.CounterVec
+	adaptTimer    *prometheus.HistogramVec
+	adaptBids     *prometheus.CounterVec
+	adaptPrices   *prometheus.HistogramVec
+	cookieSync    prometheus.Counter
+	userID        *prometheus.CounterVec
+}
+
+// NewMetrics constructs the appropriate options for the Prometheus metrics. Needs to be fed the promethus config
+// Its own function to keep the metric creation function cleaner.
+func NewMetrics(cfg config.PrometheusMetrics) *Metrics {
+	// define the buckets for timers
+	timerBuckets := prometheus.LinearBuckets(0.05, 0.05, 20)
+	timerBuckets = append(timerBuckets, []float64{1.5, 2.0, 3.0, 5.0, 10.0, 50.0}...)
+
+	standardLabelNames := []string{"source", "type", "pubid", "browser", "cookie", "status"}
+
+	adapterLabelNames := []string{"source", "type", "pubid", "browser", "cookie", "status", "adapter"}
```

Comment:
I stumbled on this in the Prometheus docs, which is a little concerning: https://prometheus.io/docs/practices/naming/#labels

> Remember that every unique combination of key-value label pairs represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), **such as user IDs**, email addresses, or other unbounded sets of values.

The fact that they specifically include user IDs is a little telling.

Is it a breaking change to add labels later? If so, it might be a good idea to leave that one out and see how Prometheus works under "normal" circumstances. Then we can add it back in and see if it's really a problem.